src/SparkFun_u-blox_GNSS_Arduino_Library.h

```diff
@@ -1292,6 +1305,9 @@ class SFE_UBLOX_GNSS
 	uint8_t *payloadCfg = NULL;
 	uint8_t *payloadAuto = NULL;
 
+	uint8_t spiBuffer[20]; 				// A small buffer to store any bytes being recieved back from the device while we are sending via SPI
```

Comment:
I'm struggling to think through how large ```spiBuffer``` should be. Hard coding it to 20 bytes is not a good idea. As a minimum I suggest changing 20 to something like ```spiBufferSize``` and either #define it or add it as a global variable?

But surely 20 is not enough? I see that bytes only get added to the buffer by ```spiTransfer```. But ```spiTransfer``` can surely transfer much more than 20 bytes? Does this mean that ```spiBuffer``` needs to be the same size as ```packetCfgPayloadSize``` (in which case it needs dynamic memory allocation) or if it can safely be limited to the maximum SPI transfer size (which varies from platform to platform). I think I suggest using the same technique as we use for I2C. Please have a look at ```setI2CTransactionSize``` and ```getI2CTransactionSize```. I think we need ```setSPITransactionSize``` and ```getSPITransactionSize```?

Reply:
Thanks Paul - excellent comments and discussion. The spiTransfer buffer only needs to be as big as the largest message that we will SEND to the device. It's used to hold any bytes returned from the device while we are sending to it as we get bytes back only when we write to the device. What's the maximum size of a message being sent? Hmm... I've just checked - there are instances where it needs to be greater than 20. How about dynamically allocating it every time we send a command and deleting it when it has been emptied? I can implement that...

Reply:
Looks like dynamic allocation of memory is a bad idea. I've tried it on my ESP32 but I'm corrupting the heap by deleting an existing array or doing new every time:
```
// Allocate memory for the spi buffer
  spiBufferSize = outgoingUBX->len + 8;  // 8 extra bytes for headers, id, class, length and checksums
  spiBufferIndex = 0;
  if (spiBuffer == NULL) //Memory has not yet been allocated - so use new
  {
    spiBuffer = new uint8_t[spiBufferSize];
  }
  else // Memory has already been allocated, so resize
  {    
    delete[] spiBuffer;
    uint8_t *newSpiBuffer = new uint8_t[spiBufferSize];    
    newSpiBuffer = spiBuffer;
  }
  if (spiBuffer == NULL) { 
    if ((_printDebug == true) || (_printLimitedDebug == true)) // This is important. Print this if doing limited debugging
    {
      _debugSerial->print(F("process: memory allocation failed for SPI Buffer!"));      
    }
  }
```

Reply:
This causes a guru meditation error on delete[] spiBuffer. I see this kind of code used elsewhere in the lib, so wonder if it would also cause issues? Safest way is to create an array that's big enough the first time we need it. I've added a define for MAX_SPI_BUFFER_SIZE to 128. I'm sure this is big enough... Will commit pending testing.

Reply:
Oops. Sorry - I should have added a more descriptive commit message. I've been up since 4am working on this on and off! Couldn't sleep ;-)

Reply:
I suspect we might still need ```setSPITransactionSize``` (and ```getSPITransactionSize```) so the user can set the maximum SPI transfer size depending on which platform they are using. One of the things I need to do after merging is test this on a standard ATmega328 RedBoard. We do still have many users using older processors with much smaller buffer sizes; the buffers on processors like ESP32 are huge by comparison.

Reply:
Thanks Paul - is it really still a buffer for SPI? I don't think there's a buffer as such in the SPI lib? If we allocate too much space, just run the risk of running out of SRAM?

Andrew

Reply:
OK. Yes. Agreed. I was forgetting that SPI transfers are performed byte by byte; and that where you store the transferred data is up to you. Even ```SPI.transfer(buffer, size)``` relies on you having created a buffer large enough to hold and store the transferred data.
Resolved... Thank you!